File: kernel/src/main/kotlin/ninja/blacknet/transaction/CreateHTLC.kt
/*
 * Copyright (c) 2018-2020 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.transaction

import kotlinx.serialization.Serializable
import ninja.blacknet.contract.HashLock
import ninja.blacknet.contract.HashTimeLockContractId
import ninja.blacknet.contract.TimeLock
import ninja.blacknet.core.*
import ninja.blacknet.crypto.Blake2b.buildHash
import ninja.blacknet.crypto.Hash
import ninja.blacknet.crypto.PublicKey
import ninja.blacknet.crypto.encodeByteArray
import ninja.blacknet.serialization.LongSerializer

/**
 * 創建合約
 */
@Serializable
class CreateHTLC(
        @Serializable(with = LongSerializer::class)
        val amount: Long,
        val to: PublicKey,
        val timeLock: TimeLock,
        val hashLock: HashLock
) : TxData {
    fun id(hash: Hash, dataIndex: Int) = HashTimeLockContractId(
        buildHash {
            encodeByteArray(hash.bytes);
            encodeInt(dataIndex);
        }
    )

    override fun processLedgerImpl(tx: Transaction, hash: Hash, dataIndex: Int, ledger: Ledger): Status {
        try {
            timeLock.validate()
        } catch (e: Throwable) {
            return Invalid("Invalid time lock ${e.message}")
        }
        try {
            hashLock.validate()
        } catch (e: Throwable) {
            return Invalid("Invalid hash lock ${e.message}")
        }

        if (amount == 0L) {
            return Invalid("Invalid amount")
        }

        val account = ledger.getAccount(tx.from)!!
        val status = account.credit(amount)
        if (status != Accepted) {
            return status
        }

        val id = id(hash, dataIndex)
        val htlc = HTLC(ledger.height(), ledger.blockTime(), amount, tx.from, to, timeLock, hashLock)
        ledger.setAccount(tx.from, account)
        ledger.addHTLC(id, htlc)
        return Accepted
    }

    fun involves(publicKey: PublicKey) = to == publicKey
}

File: kernel/src/main/rust/transaction/createhtlc.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::amount::Amount;
use crate::blake2b::{Blake2b256, Hash};
use crate::ed25519::PublicKey;
use crate::error::{Error, Result};
use crate::hashlock::HashLock;
use crate::htlc::HTLC;
use crate::timelock::TimeLock;
use crate::transaction::{CoinTx, Transaction, TxData};
use alloc::borrow::ToOwned;
use digest::Digest;
use serde::{Deserialize, Serialize};

pub type HashTimeLockContractId = [u8; 32];

fn id(hash: Hash, data_index: u32) -> HashTimeLockContractId {
    let mut hasher = Blake2b256::new();
    hasher.update(hash);
    hasher.update(data_index.to_be_bytes());
    hasher.finalize().into()
}

#[derive(Deserialize, Serialize)]
pub struct CreateHTLC {
    amount: Amount,
    to: PublicKey,
    time_lock: TimeLock,
    hash_lock: HashLock,
}

impl TxData for CreateHTLC {
    fn process_impl(
        &self,
        tx: Transaction,
        hash: Hash,
        data_index: u32,
        coin_tx: impl CoinTx,
    ) -> Result<()> {
        self.time_lock.validate()?;
        self.hash_lock.validate()?;

        if self.amount == Amount::ZERO {
            return Err(Error::Invalid("Invalid amount".to_owned()));
        }

        let mut account = coin_tx.get_account(tx.from)?;
        account.credit(self.amount)?;

        let id = id(hash, data_index);
        let htlc = HTLC {
            height: coin_tx.height(),
            time: coin_tx.block_time(),
            amount: self.amount,
            from: tx.from,
            to: self.to,
            time_lock: self.time_lock.clone(),
            hash_lock: self.hash_lock.clone(),
        };
        coin_tx.set_account(tx.from, account);
        coin_tx.add_htlc(id, htlc);
        Ok(())
    }
}
